Lib/functools.py

```
@@ -1026,8 +1026,8 @@ def __init__(self, func):
         self.dispatcher = singledispatch(func)
         self.func = func
 
-        import weakref # see comment in singledispatch function
-        self._method_cache = weakref.WeakKeyDictionary()
+    def __set_name__(self, obj, name):
+        self.attrname = name
```

Check `cached_property.__set_name__`, it has some more stuff in it - might be needed here as well.

---

Hmm. The additions there prevent something like this:
```
@dataclass(frozen=True)
class A:
    value: int

    @singledispatchmethod
    def dispatch(self, x):
        return id(self)

    renamed_dispatch = dispatch # allowed? if so, how should it behave
```
The corresponding test for the `cached_property` for this is https://github.com/python/cpython/blob/34e840f9ddd9c04991cf004e2594c6a1e0e278d3/Lib/test/test_functools.py#L3315.

But on current main renaming _is_ allowed for the singledispatchmethod. 

I am not sure here what the desired behavior is (and why)

---

If this implementation is desirable, maybe later someone who knows more about this can comment.

---

As far as I know, the only reason cached properties can't be renamed is because the cache is keyed by the attribute's name.

Allowing a rebind would disconnect the cached property from it's cached value.

---

Actually, I think you might want to something along these lines (ignoring error handling):
```python
    if self.attrname:
        cache[name] = cache.pop(self.attrname)
    self.attrname = name

---

> Allowing a rebind would disconnect the cached property from it's cached value.

This is kind of the same situation.

If rename is allowed, then it would simply cache to the last `attrname`. Drawback is that there is a small risk for unused cached methods.

I think it might be most straight forward to copy+paste `cached_property.__set_name__`. It does seem a sensible restriction. It comes at expense of flexibility, but personally, I have never run into *that* `TypeError`.

Also, it will be easier to address changes/improvements when 2 implementations that use the same caching approach are aligned.